_WKHitTestResult should expose a linkLocalDataMIMEType property

jeffmapple opened the pull request:
#### 00a7a1e70c1a109781ddb1547e6e58c5e706211e
<pre>
_WKHitTestResult should expose a linkLocalDataMIMEType property
<a href="https://bugs.webkit.org/show_bug.cgi?id=266730">https://bugs.webkit.org/show_bug.cgi?id=266730</a>
<a href="https://rdar.apple.com/119404077">rdar://119404077</a>

Reviewed by Aditya Keerthi.

Add linkLocalDataMIMEType to WebHitTestResultData (and the underlying plumbing to support calculating
this string) and expose it via a property on _WKHitTestResult.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::hasLocalDataForLinkURL const):
Added.

* Source/WebCore/rendering/HitTestResult.h:
Added hasLocalDataForLinkURL() and exported it so WebKit can use it.

* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
Added linkLocalDataMIMEType property.

* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult linkLocalDataMIMEType]):
Added.

* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::webFrameFromHitTestResult):
Refactored from frameInfoDataFromHitTestResult() so we can reuse this logic.

(WebKit::linkLocalDataMIMETypeFromHitTestResult):
Added.

(WebKit::WebHitTestResultData::WebHitTestResultData):
Initialize linkLocalDataMIMEType.

(WebKit::WebHitTestResultData::frameInfoDataFromHitTestResult):
Refactor some logic into webFrameFromHitTestResult().

* Source/WebKit/Shared/WebHitTestResultData.h:
Added linkLocalDataMIMEType.

* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
Added linkLocalDataMIMEType.

* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::linkLocalDataMIMEType const):
Added.

* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added HitTestResultLinkLocalDataMIMEType, loads an image inside a link to itself and verifies that the
linkLocalDataMIMEType is &quot;image/jpeg&quot;.

Canonical link: <a href="https://commits.webkit.org/272600@main">https://commits.webkit.org/272600@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b6ae2321947cb3ca739bb280b12dc00ec68793a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34092 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34832 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29235 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33126 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13363 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8220 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28793 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32694 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9278 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8082 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8227 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28776 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36175 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29341 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29215 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34351 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6294 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32216 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10009 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7535 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8992 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8909 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
```diff
@@ -45,6 +45,7 @@ WK_CLASS_AVAILABLE(macos(10.12), ios(16.0))
 
 @property (nonatomic, readonly, copy) NSURL *absolutePDFURL;
 @property (nonatomic, readonly, copy) NSURL *absoluteLinkURL;
+@property (nonatomic, readonly, copy) NSString *linkDataMIMEType WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
Ideally, the name of the property would also indicate that it only applies for local data.

Maybe `linkLocalDataMIMEType`?

pxlcoder commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -671,6 +671,18 @@ URL HitTestResult::absoluteLinkURL() const
 
+bool HitTestResult::hasLocalDataForLinkURL() const
+{
+    URL linkURL = absoluteLinkURL();
```
Nit:
```suggestion
    auto linkURL = absoluteLinkURL();
```

pxlcoder commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -671,6 +671,18 @@ URL HitTestResult::absoluteLinkURL() const
+    if (linkURL.isEmpty())
+        return false;
+
+    if (auto* page = m_innerURLElement->document().page())
```
As we are starting to adopt smart pointers everywhere:
```suggestion
    if (RefPtr page = m_innerURLElement->document().page())
```

pxlcoder commented on Source/WebKit/Shared/WebHitTestResultData.cpp:
```diff
@@ -47,6 +47,27 @@ static WebHitTestResultData::ElementType elementTypeFromHitTestResult(const HitT
+
+    RefPtr webFrame = webFrameFromHitTestResult(hitTestResult);
+    if (!webFrame)
+        return { };
```
```suggestion
        return nullString();
```

pxlcoder commented on Source/WebKit/Shared/WebHitTestResultData.cpp:
```diff
@@ -47,6 +47,27 @@ static WebHitTestResultData::ElementType elementTypeFromHitTestResult(const HitT
+static String linkDataMIMETypeFromHitTestResult(const HitTestResult& hitTestResult)
+{
+    if (!hitTestResult.hasLocalDataForLinkURL())
+        return { };
```
```suggestion
        return nullString();
```

jeffmapple committed:
_WKHitTestResult should expose a linkLocalDataMIMEType property
https://bugs.webkit.org/show_bug.cgi?id=266730
rdar://119404077

Reviewed by Aditya Keerthi.

Add linkLocalDataMIMEType to WebHitTestResultData (and the underlying plumbing to support calculating
this string) and expose it via a property on _WKHitTestResult.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::hasLocalDataForLinkURL const):
Added.

* Source/WebCore/rendering/HitTestResult.h:
Added hasLocalDataForLinkURL() and exported it so WebKit can use it.

* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
Added linkLocalDataMIMEType property.

* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult linkLocalDataMIMEType]):
Added.

* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::webFrameFromHitTestResult):
Refactored from frameInfoDataFromHitTestResult() so we can reuse this logic.

(WebKit::linkLocalDataMIMETypeFromHitTestResult):
Added.

(WebKit::WebHitTestResultData::WebHitTestResultData):
Initialize linkLocalDataMIMEType.

(WebKit::WebHitTestResultData::frameInfoDataFromHitTestResult):
Refactor some logic into webFrameFromHitTestResult().

* Source/WebKit/Shared/WebHitTestResultData.h:
Added linkLocalDataMIMEType.

* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
Added linkLocalDataMIMEType.

* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::linkLocalDataMIMEType const):
Added.

* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added HitTestResultLinkLocalDataMIMEType, loads an image inside a link to itself and verifies that the
linkLocalDataMIMEType is "image/jpeg".

Canonical link: https://commits.webkit.org/272600@main
